### PR TITLE
Fix bug in async TP handling of "reshape -> scaled mm -> reshape" pattern for float8 row-wise scaling

### PR DESCRIPTION
**Pull request by danielvegamyhre**
Part of https://github.com/pytorch/torchtitan/issues/864

## Summary 
While testing torchtitan with float8 training with rowwise scaling + async TP, a [bug](https://github.com/pytorch/torchtitan/issues/864) was discovered. The symptom was the scaling factor dims did not match the dims of the tensor the scales were to be applied to.

My [root cause analysis](https://github.com/pytorch/torchtitan/issues/864#issuecomment-2672465060) determined the reason is that when async TP graph manipulation constructs the `fused_scaled_matmul_reduce_scatter` op, it does not yet handle the "reshape -> scaled mm -> reshape" pattern used in torchao [here](https://github.com/pytorch/ao/blob/ed361ff5c7dd33aba9b4a0da2bd744de5a5debfb/torchao/float8/float8_linear.py#L122-L124) - specifically when row-wise scales are being used.

## TL;DR of root cause
- When a Float8Tensor is reshaped, the scale is reshaped along with it so the dimensions are aligned.
- In the graph manipulation logic of the micropipeline TP post grad pass, the scaled_mm `A tensor` node is referencing the tensor _before_ to the reshape op, but referencing the `A_scale` node _after_ the reshape op.
- To solve this, if a reshape -> scaled mm -> reshape pattern is detected, we can ensure both the tensor and scale used are from _before_ the reshape.

**Note:** the reason we don't use the tensor and scale from _after_  the reshape is because then the `scatter_dim`, which corresponds to the original tensor shape, would now be outdated, and keeping the scatter dim in sync with arbitrary reshapes would be complicated/not feasible. Furthermore, using the tensor / scale from before the reshape ensures the `fused_scaled_matmul_reduce_scatter` keeps the intended `(a,b,c) @ (c,d) = (a,b,d)` shape sequence

## Test plan
- Added new unit tests ensuring "reshape -> scaled mm -> reshape" pattern with row-wise scales is supported.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

**Comment by danielvegamyhre** [comment]
Strange, the failing test is passing locally:

```bash
[danvm@devgpu006.vll6 /data/users/danvm/pytorch (async-tp)]$ python test/distributed/tensor/parallel/test_micro_pipeline_tp.py MicroPipelineTPTest.test_fuse_scaled_matmul_reduce_scatter_A_dims_3_scatter_dim_2
.
----------------------------------------------------------------------
Ran 1 test in 1.866s

OK
```

File: torch/_inductor/fx_passes/micro_pipeline_tp.py [hunk 1]
```diff
@@ -374,23 +374,46 @@ def __post_init__(self):
...
 
         def get_arg(node: torch.fx.Node, idx: int, default: Any) -> Any:
             if idx >= len(node.args):
                 return default
             return node.args[idx]
 
+        is_reshape_mm_reshape_pattern = match[0].target == aten.reshape.default
+        mm_node = match[1] if is_reshape_mm_reshape_pattern else match[0]
+
+        # case 1: single node match (mm):
```
Comment (vkuzo) [inline torch/_inductor/fx_passes/micro_pipeline_tp.py:398]: just curious, do we know who is using case 1 and how are they handling incoming shapes of rank 3+?
